File: vscode-extension/src/test/protocol.test.ts
import * as assert from "assert";
import {
  Event,
  Command,
  parseEvent,
  serializeCommand,
  isFrame,
  isStreamStarted,
  isStreamStopped,
  isStreamStatus,
} from "../protocol";

suite("Protocol", () => {
  suite("Event parsing", () => {
    test("parses Frame event", () => {
      const json = '{"streamId":"a","frame":{"device":"iPhone 16 Pro","file":"HogeView.swift","data":"base64data"}}';
      const event = parseEvent(json);
      assert.ok(event);
      assert.strictEqual(event.streamId, "a");
      assert.ok(event.frame);
      assert.strictEqual(event.frame.device, "iPhone 16 Pro");
      assert.strictEqual(event.frame.file, "HogeView.swift");
      assert.strictEqual(event.frame.data, "base64data");
    });

    test("parses StreamStarted event", () => {
      const json = '{"streamId":"b","streamStarted":{"previewCount":3}}';
      const event = parseEvent(json);
      assert.ok(event);
      assert.strictEqual(event.streamId, "b");
      assert.ok(event.streamStarted);
      assert.strictEqual(event.streamStarted.previewCount, 3);
    });

    test("parses StreamStopped event", () => {
      const json = '{"streamId":"c","streamStopped":{"reason":"build_error","message":"Build failed","diagnostic":"error: expected }"}}';
      const event = parseEvent(json);
      assert.ok(event);
      assert.strictEqual(event.streamId, "c");
      assert.ok(event.streamStopped);
      assert.strictEqual(event.streamStopped.reason, "build_error");
      assert.strictEqual(event.streamStopped.message, "Build failed");
    });

    test("parses StreamStatus event", () => {
      const json = '{"streamId":"d","streamStatus":{"phase":"building"}}';
      const event = parseEvent(json);
      assert.ok(event);
      assert.strictEqual(event.streamId, "d");
      assert.ok(event.streamStatus);
      assert.strictEqual(event.streamStatus.phase, "building");
    });

    test("returns undefined for invalid JSON", () => {
      assert.strictEqual(parseEvent("not json"), undefined);
      assert.strictEqual(parseEvent(""), undefined);
      assert.strictEqual(parseEvent("{truncated"), undefined);
    });

    test("returns undefined for non-object JSON", () => {
      assert.strictEqual(parseEvent("[1,2,3]"), undefined);
      assert.strictEqual(parseEvent('"string"'), undefined);
      assert.strictEqual(parseEvent("42"), undefined);
      assert.strictEqual(parseEvent("null"), undefined);
    });

    test("returns undefined when streamId is missing", () => {
      assert.strictEqual(parseEvent('{"frame":{"data":"abc"}}'), undefined);
    });

    test("returns undefined when streamId is not a string", () => {
      assert.strictEqual(parseEvent('{"streamId":123,"frame":{"data":"abc"}}'), undefined);
    });

    test("tolerates unknown fields", () => {
      const json = '{"streamId":"a","frame":{"device":"iPhone","file":"V.swift","data":"abc","extra":"ignored"},"futureField":{}}';
      const event = parseEvent(json);
      assert.ok(event);
      assert.strictEqual(event.streamId, "a");
      assert.ok(event.frame);
      assert.strictEqual(event.frame.data, "abc");
    });
  });

  suite("Command serialization", () => {
    test("serializes AddStream command", () => {
      const cmd: Command = {
        streamId: "s1",
        addStream: {
          file: "/path/to/View.swift",
          deviceType: "com.apple.CoreSimulator.SimDeviceType.iPhone-16-Pro",
          runtime: "com.apple.CoreSimulator.SimRuntime.iOS-18-2",
        },
      };
      const json = serializeCommand(cmd);
      const parsed = JSON.parse(json);
      assert.strictEqual(parsed.streamId, "s1");
      assert.strictEqual(parsed.addStream.file, "/path/to/View.swift");
      assert.strictEqual(parsed.addStream.deviceType, "com.apple.CoreSimulator.SimDeviceType.iPhone-16-Pro");
      assert.strictEqual(parsed.addStream.runtime, "com.apple.CoreSimulator.SimRuntime.iOS-18-2");
    });

    test("serializes RemoveStream command", () => {
      const cmd: Command = { streamId: "s1", removeStream: {} };
      const json = serializeCommand(cmd);
      const parsed = JSON.parse(json);
      assert.strictEqual(parsed.streamId, "s1");
      assert.deepStrictEqual(parsed.removeStream, {});
    });

    test("serializes SwitchFile command", () => {
      const cmd: Command = { streamId: "s1", switchFile: { file: "/path/to/Other.swift" } };
      const json = serializeCommand(cmd);
      const parsed = JSON.parse(json);
      assert.strictEqual(parsed.switchFile.file, "/path/to/Other.swift");
    });

    test("serializes NextPreview command", () => {
      const cmd: Command = { streamId: "s1", nextPreview: {} };
      const json = serializeCommand(cmd);
      const parsed = JSON.parse(json);
      assert.strictEqual(parsed.streamId, "s1");
      assert.deepStrictEqual(parsed.nextPreview, {});
    });

    test("serializes Input command with TouchDown", () => {
      const cmd: Command = {
        streamId: "s1",
        input: { touchDown: { x: 0.5, y: 0.3 } },
      };
      const json = serializeCommand(cmd);
      const parsed = JSON.parse(json);
      assert.strictEqual(parsed.input.touchDown.x, 0.5);
      assert.strictEqual(parsed.input.touchDown.y, 0.3);
    });

    test("serializes Input command with Text", () => {
      const cmd: Command = {
        streamId: "s1",
        input: { text: { value: "a" } },
      };
      const json = serializeCommand(cmd);
      const parsed = JSON.parse(json);
      assert.strictEqual(parsed.input.text.value, "a");
    });

    test("omits undefined payload fields", () => {
      const cmd: Command = {
        streamId: "s1",
        addStream: { file: "/p", deviceType: "dt", runtime: "rt" },
      };
      const json = serializeCommand(cmd);
      assert.ok(!json.includes("removeStream"));
      assert.ok(!json.includes("switchFile"));
      assert.ok(!json.includes("nextPreview"));
      assert.ok(!json.includes("input"));
    });

    test("round-trip: serialize then parse as Event-like JSON", () => {
      // This simulates the CLI receiving a serialized Command
      const cmd: Command = {
        streamId: "s1",
        addStream: { file: "/p", deviceType: "dt", runtime: "rt" },
      };
      const json = serializeCommand(cmd);
      const parsed = JSON.parse(json);
      assert.strictEqual(parsed.streamId, "s1");
      assert.strictEqual(parsed.addStream.file, "/p");
    });
  });

  suite("Type guards", () => {
    test("isFrame returns true for Frame events", () => {
      const event: Event = {
        streamId: "a",
        frame: { device: "iPhone", file: "V.swift", data: "abc" },
      };
      assert.strictEqual(isFrame(event), true);
      assert.strictEqual(isStreamStarted(event), false);
      assert.strictEqual(isStreamStopped(event), false);
      assert.strictEqual(isStreamStatus(event), false);
    });

    test("isStreamStarted returns true for StreamStarted events", () => {
      const event: Event = {
        streamId: "a",
        streamStarted: { previewCount: 2 },
      };
      assert.strictEqual(isFrame(event), false);
      assert.strictEqual(isStreamStarted(event), true);
    });

    test("isStreamStopped returns true for StreamStopped events", () => {
      const event: Event = {
        streamId: "a",
        streamStopped: { reason: "removed", message: "", diagnostic: "" },
      };
      assert.strictEqual(isStreamStopped(event), true);
      assert.strictEqual(isFrame(event), false);
    });

    test("isStreamStatus returns true for StreamStatus events", () => {
      const event: Event = {
        streamId: "a",
        streamStatus: { phase: "building" },
      };
      assert.strictEqual(isStreamStatus(event), true);
      assert.strictEqual(isFrame(event), false);
    });

    test("all guards return false for empty event", () => {
      const event: Event = { streamId: "a" };
      assert.strictEqual(isFrame(event), false);
      assert.strictEqual(isStreamStarted(event), false);
      assert.strictEqual(isStreamStopped(event), false);
      assert.strictEqual(isStreamStatus(event), false);
    });
  });

  suite("Cross-language compatibility", () => {
    test("Go-produced Frame JSON parses correctly in TS", () => {
      // Simulate JSON output from Go's json.Marshal of Event{StreamID:"a",Frame:&Frame{...}}
      const goJSON = '{"streamId":"a","frame":{"device":"iPhone 16 Pro","file":"HogeView.swift","data":"AAAA"}}';
      const event = parseEvent(goJSON);
      assert.ok(event);
      assert.ok(isFrame(event));
      assert.strictEqual(event.frame!.device, "iPhone 16 Pro");
      assert.strictEqual(event.frame!.data, "AAAA");
    });

    test("TS-produced AddStream JSON matches Go expected format", () => {
      const cmd: Command = {
        streamId: "stream-1",
        addStream: {
          file: "/path/to/HogeView.swift",
          deviceType: "com.apple.CoreSimulator.SimDeviceType.iPhone-16-Pro",
          runtime: "com.apple.CoreSimulator.SimRuntime.iOS-18-2",
        },
      };
      const json = serializeCommand(cmd);
      const parsed = JSON.parse(json);

      // Verify field names match Go json tags (camelCase)
      assert.ok("streamId" in parsed);
      assert.ok("addStream" in parsed);
      assert.ok("file" in parsed.addStream);
      assert.ok("deviceType" in parsed.addStream);
      assert.ok("runtime" in parsed.addStream);
    });
  });
});

File: cmd/internal/preview/protocol_test.go
package preview

import (
	"encoding/json"
	"strings"
	"testing"

	pb "github.com/k-kohey/axe/internal/preview/previewproto"
)

func TestEvent_MarshalRoundTrip(t *testing.T) {
	tests := []struct {
		name  string
		event *pb.Event
	}{
		{
			name: "Frame",
			event: &pb.Event{
				StreamId: "abc-123",
				Payload:  &pb.Event_Frame{Frame: &pb.Frame{Device: "iPhone 16 Pro", File: "ContentView.swift", Data: "base64data"}},
			},
		},
		{
			name: "StreamStarted",
			event: &pb.Event{
				StreamId: "abc-123",
				Payload:  &pb.Event_StreamStarted{StreamStarted: &pb.StreamStarted{PreviewCount: 3}},
			},
		},
		{
			name: "StreamStopped",
			event: &pb.Event{
				StreamId: "abc-123",
				Payload:  &pb.Event_StreamStopped{StreamStopped: &pb.StreamStopped{Reason: "build_error", Message: "Build failed", Diagnostic: "error: expected '}'"}},
			},
		},
		{
			name: "StreamStatus",
			event: &pb.Event{
				StreamId: "abc-123",
				Payload:  &pb.Event_StreamStatus{StreamStatus: &pb.StreamStatus{Phase: "building"}},
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			data, err := MarshalEvent(tt.event)
			if err != nil {
				t.Fatalf("Marshal failed: %v", err)
			}

			// Verify the output is valid JSON.
			var raw map[string]any
			if err := json.Unmarshal(data, &raw); err != nil {
				t.Fatalf("protojson output is not valid JSON: %v", err)
			}

			// Verify streamId is present.
			if raw["streamId"] != "abc-123" {
				t.Errorf("expected streamId=abc-123, got %v", raw["streamId"])
			}
		})
	}
}

func TestCommand_UnmarshalRoundTrip(t *testing.T) {
	tests := []struct {
		name    string
		json    string
		checkFn func(t *testing.T, cmd *pb.Command)
	}{
		{
			name: "AddStream",
			json: `{"streamId":"stream-1","addStream":{"file":"/path/to/View.swift","deviceType":"com.apple.CoreSimulator.SimDeviceType.iPhone-16-Pro","runtime":"com.apple.CoreSimulator.SimRuntime.iOS-18-2"}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetStreamId() != "stream-1" {
					t.Errorf("StreamId = %q, want %q", cmd.GetStreamId(), "stream-1")
				}
				if cmd.GetAddStream() == nil {
					t.Fatal("expected AddStream to be non-nil")
				}
				if cmd.GetAddStream().GetFile() != "/path/to/View.swift" {
					t.Errorf("File = %q, want %q", cmd.GetAddStream().GetFile(), "/path/to/View.swift")
				}
			},
		},
		{
			name: "RemoveStream",
			json: `{"streamId":"stream-1","removeStream":{}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetRemoveStream() == nil {
					t.Fatal("expected RemoveStream to be non-nil")
				}
			},
		},
		{
			name: "SwitchFile",
			json: `{"streamId":"stream-1","switchFile":{"file":"/path/to/Other.swift"}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetSwitchFile() == nil {
					t.Fatal("expected SwitchFile to be non-nil")
				}
				if cmd.GetSwitchFile().GetFile() != "/path/to/Other.swift" {
					t.Errorf("File = %q, want %q", cmd.GetSwitchFile().GetFile(), "/path/to/Other.swift")
				}
			},
		},
		{
			name: "NextPreview",
			json: `{"streamId":"stream-1","nextPreview":{}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetNextPreview() == nil {
					t.Fatal("expected NextPreview to be non-nil")
				}
			},
		},
		{
			name: "Input_TouchDown",
			json: `{"streamId":"stream-1","input":{"touchDown":{"x":0.5,"y":0.3}}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetInput() == nil {
					t.Fatal("expected Input to be non-nil")
				}
				if cmd.GetInput().GetTouchDown() == nil {
					t.Fatal("expected TouchDown to be non-nil")
				}
				if cmd.GetInput().GetTouchDown().GetX() != 0.5 {
					t.Errorf("X = %v, want 0.5", cmd.GetInput().GetTouchDown().GetX())
				}
			},
		},
		{
			name: "Input_Text",
			json: `{"streamId":"stream-1","input":{"text":{"value":"a"}}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetInput() == nil {
					t.Fatal("expected Input to be non-nil")
				}
				if cmd.GetInput().GetText() == nil {
					t.Fatal("expected Text to be non-nil")
				}
				if cmd.GetInput().GetText().GetValue() != "a" {
					t.Errorf("Value = %q, want %q", cmd.GetInput().GetText().GetValue(), "a")
				}
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cmd, err := UnmarshalCommand([]byte(tt.json))
			if err != nil {
				t.Fatalf("UnmarshalCommand failed: %v", err)
			}
			tt.checkFn(t, cmd)
		})
	}
}

func TestEvent_OmitEmpty(t *testing.T) {
	// Only Frame is set; other payload fields should be omitted.
	e := &pb.Event{
		StreamId: "s1",
		Payload:  &pb.Event_Frame{Frame: &pb.Frame{Device: "iPhone 16 Pro", File: "View.swift", Data: "abc"}},
	}

	data, err := MarshalEvent(e)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	s := string(data)
	for _, key := range []string{`"streamStarted"`, `"streamStopped"`, `"streamStatus"`} {
		if strings.Contains(s, key) {
			t.Errorf("expected %s to be omitted, got: %s", key, s)
		}
	}

	// Verify streamId and frame are present.
	if !strings.Contains(s, `"streamId"`) {
		t.Errorf("expected streamId to be present, got: %s", s)
	}
	if !strings.Contains(s, `"frame"`) {
		t.Errorf("expected frame to be present, got: %s", s)
	}
}

func TestCommand_UnknownFieldTolerance(t *testing.T) {
	// Simulate a future extension sending extra fields.
	jsonStr := `{"streamId":"s1","addStream":{"file":"/path","deviceType":"dt","runtime":"rt","unknownField":"value"},"futureField":42}`

	cmd, err := UnmarshalCommand([]byte(jsonStr))
	if err != nil {
		t.Fatalf("Unmarshal with unknown fields should succeed: %v", err)
	}

	if cmd.GetStreamId() != "s1" {
		t.Errorf("StreamId = %q, want %q", cmd.GetStreamId(), "s1")
	}
	if cmd.GetAddStream() == nil {
		t.Fatal("AddStream should not be nil")
	}
	if cmd.GetAddStream().GetFile() != "/path" {
		t.Errorf("AddStream.File = %q, want %q", cmd.GetAddStream().GetFile(), "/path")
	}
}

func TestCommand_InvalidJSON(t *testing.T) {
	tests := []struct {
		name  string
		input string
	}{
		{"empty string", ""},
		{"not JSON", "hello"},
		{"truncated", `{"streamId":"s1"`},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := UnmarshalCommand([]byte(tt.input))
			if err == nil {
				t.Errorf("expected error for input %q, got nil", tt.input)
			}
		})
	}
}

func TestEvent_JSONFieldNames(t *testing.T) {
	// Verify that JSON uses camelCase field names (matching the design doc).
	e := &pb.Event{
		StreamId: "s1",
		Payload:  &pb.Event_StreamStarted{StreamStarted: &pb.StreamStarted{PreviewCount: 2}},
	}

	data, err := MarshalEvent(e)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	s := string(data)
	// Should use camelCase
	if !strings.Contains(s, `"streamId"`) {
		t.Errorf("expected camelCase streamId, got: %s", s)
	}
	if !strings.Contains(s, `"streamStarted"`) {
		t.Errorf("expected camelCase streamStarted, got: %s", s)
	}
	if !strings.Contains(s, `"previewCount"`) {
		t.Errorf("expected camelCase previewCount, got: %s", s)
	}
	// Should NOT use snake_case
	if strings.Contains(s, `"stream_id"`) {
		t.Errorf("unexpected snake_case stream_id, got: %s", s)
	}
}

func TestCommand_JSONFieldNames(t *testing.T) {
	// Verify AddStream JSON from TS extension is correctly parsed.
	jsonStr := `{"streamId":"s1","addStream":{"file":"/path","deviceType":"dt","runtime":"rt"}}`
	cmd, err := UnmarshalCommand([]byte(jsonStr))
	if err != nil {
		t.Fatalf("UnmarshalCommand failed: %v", err)
	}
	if cmd.GetStreamId() != "s1" {
		t.Errorf("expected streamId = s1, got %s", cmd.GetStreamId())
	}
	if cmd.GetAddStream() == nil {
		t.Fatal("expected addStream to be non-nil")
	}
	if cmd.GetAddStream().GetDeviceType() != "dt" {
		t.Errorf("expected deviceType = dt, got %s", cmd.GetAddStream().GetDeviceType())
	}
}

func TestInput_UnmarshalRoundTrip(t *testing.T) {
	tests := []struct {
		name    string
		json    string
		checkFn func(t *testing.T, cmd *pb.Command)
	}{
		{
			name: "TouchDown",
			json: `{"streamId":"s1","input":{"touchDown":{"x":0.1,"y":0.2}}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetInput().GetTouchDown() == nil {
					t.Fatal("expected TouchDown")
				}
				if cmd.GetInput().GetTouchDown().GetX() != 0.1 {
					t.Errorf("X = %v, want 0.1", cmd.GetInput().GetTouchDown().GetX())
				}
			},
		},
		{
			name: "TouchMove",
			json: `{"streamId":"s1","input":{"touchMove":{"x":0.3,"y":0.4}}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetInput().GetTouchMove() == nil {
					t.Fatal("expected TouchMove")
				}
			},
		},
		{
			name: "TouchUp",
			json: `{"streamId":"s1","input":{"touchUp":{"x":0.5,"y":0.6}}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetInput().GetTouchUp() == nil {
					t.Fatal("expected TouchUp")
				}
			},
		},
		{
			name: "Text",
			json: `{"streamId":"s1","input":{"text":{"value":"x"}}}`,
			checkFn: func(t *testing.T, cmd *pb.Command) {
				if cmd.GetInput().GetText() == nil {
					t.Fatal("expected Text")
				}
				if cmd.GetInput().GetText().GetValue() != "x" {
					t.Errorf("Value = %q, want %q", cmd.GetInput().GetText().GetValue(), "x")
				}
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cmd, err := UnmarshalCommand([]byte(tt.json))
			if err != nil {
				t.Fatalf("Unmarshal failed: %v", err)
			}
			tt.checkFn(t, cmd)
		})
	}
}

func TestStreamStopped_EmptyDiagnostic(t *testing.T) {
	// When diagnostic is empty, protojson omits it. Verify the reason field is still present.
	e := &pb.Event{
		StreamId: "s1",
		Payload:  &pb.Event_StreamStopped{StreamStopped: &pb.StreamStopped{Reason: "removed"}},
	}

	data, err := MarshalEvent(e)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	s := string(data)
	if !strings.Contains(s, `"reason"`) {
		t.Errorf("expected reason to be present, got: %s", s)
	}

	// Verify round-trip via JSON.parse on TS side would work.
	var raw map[string]any
	if err := json.Unmarshal(data, &raw); err != nil {
		t.Fatalf("JSON parse failed: %v", err)
	}
	ss, ok := raw["streamStopped"].(map[string]any)
	if !ok {
		t.Fatal("expected streamStopped to be present")
	}
	if ss["reason"] != "removed" {
		t.Errorf("reason = %v, want %q", ss["reason"], "removed")
	}
}

// TestEvent_CrossLanguageWireFormat verifies that protojson output is compatible
// with the TS extension's JSON.parse.
func TestEvent_CrossLanguageWireFormat(t *testing.T) {
	e := &pb.Event{
		StreamId: "s1",
		Payload:  &pb.Event_Frame{Frame: &pb.Frame{Device: "iPhone 16 Pro", File: "HogeView.swift", Data: "AAAA"}},
	}

	data, err := MarshalEvent(e)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	// Parse as generic JSON (simulating TS JSON.parse).
	var raw map[string]any
	if err := json.Unmarshal(data, &raw); err != nil {
		t.Fatalf("JSON.parse equivalent failed: %v", err)
	}

	if raw["streamId"] != "s1" {
		t.Errorf("streamId = %v, want %q", raw["streamId"], "s1")
	}
	frame, ok := raw["frame"].(map[string]any)
	if !ok {
		t.Fatal("expected frame to be present")
	}
	if frame["device"] != "iPhone 16 Pro" {
		t.Errorf("frame.device = %v, want %q", frame["device"], "iPhone 16 Pro")
	}
	if frame["data"] != "AAAA" {
		t.Errorf("frame.data = %v, want %q", frame["data"], "AAAA")
	}
}
